opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py

```diff
@@ -349,6 +362,326 @@ def collect(
+
+        self._count += 1
+
+        if value == 0:
+            self._zero_count += 1
+            return
+
+        self._sum += value
+
+        # 1. Use the positive buckets for positive values and the negative
+        # buckets for negative values.
+        if value > 0:
+            buckets = self._positive
+
+        else:
+            # Both exponential and logarithm mappings use only positive values
+            # so the absolute value is used here.
+            value = -value
+            buckets = self._negative
+
+        # 2. Compute the index for the value at the current scale.
+        index = self._mapping.map_to_index(value)
+
+        # 3. Determine if a change of scale is needed.
+        is_rescaling_needed = False
+
+        if len(buckets) == 0:
+            buckets._index_start = index
+            buckets._index_end = index
+            buckets._index_base = index
+
+        elif (
+            index < buckets._index_start
+            and (buckets._index_end - index) >= self._max_size
+        ):
+            is_rescaling_needed = True
+            low = index
+            high = buckets._index_end
+
+        elif (
+            index > buckets._index_end
+            and (index - buckets._index_start) >= self._max_size
+        ):
+            is_rescaling_needed = True
+            low = buckets._index_start
+            high = index
+
+        # 4. Rescale the mapping if needed.
+        if is_rescaling_needed:
+
+            change = 0
+
+            while high - low >= self._max_size:
+                high = high >> 1
+                low = low >> 1
+
+                change += 1
+
+            new_scale = self._mapping.scale - change
+
+            self._positive.downscale(change)
+            self._negative.downscale(change)
+
+            if new_scale <= 0:
+                mapping = ExponentMapping(new_scale)
+            else:
+                mapping = LogarithmMapping(new_scale)
+
+            self._mapping = mapping
+
+            index = self._mapping.map_to_index(value)
+
+        # 5. If the index is outside
+        # [buckets._index_start, buckets._index_end] readjust the buckets
+        # boundaries or add more buckets.
+        if index < buckets._index_start:
+            span = buckets._index_end - index
+
+            if span >= len(buckets.counts):
+                buckets.grow(span + 1, self._max_size)
+
+            buckets._index_start = index
+
+        elif index > buckets._index_end:
+            span = index - buckets._index_start
+
+            if span >= len(buckets.counts):
+                buckets.grow(span + 1, self._max_size)
+
+            buckets._index_end = index
+
+        # 6. Compute the index of the bucket to be incremented.
+        bucket_index = index - buckets._index_base
+
+        if bucket_index < 0:
+            bucket_index += len(buckets.counts)
+
+        # 7. Increment the bucket.
+        buckets.increment_bucket(bucket_index)
+
+    def collect(
```

Comment:
This doesn't look thread safe. IIRC, nothing is preventing new calls to `aggregate()` from coming in during collection.

It looks like we may have the same issues in ExplicitBucketHistogramAggregation and SumAggregation, maybe im misremembering if we actually do hold a lock in the code that calls this?

Reply:
Added a lock for the whole method.